src/test/test-fs-util.c

```diff
@@ -1011,6 +1011,21 @@ TEST(open_mkdir_at) {
         assert_se(subsubdir_fd >= 0);
 }
 
+TEST(open_mkdir_at_parents) {
+        _cleanup_(rm_rf_physical_and_freep) char *t = NULL;
+        _cleanup_close_ int dfd = -EBADF, tfd = -EBADF;
+
+        assert_se((tfd = mkdtemp_open(NULL, 0, &t)) >= 0);
+
+        assert_se((dfd = open_mkdir_at_p(tfd, "a/b/c", O_CLOEXEC, 0700)) >= 0);
+        assert_se(faccessat(tfd, "a/b/c", F_OK, 0) >= 0);
+        dfd = safe_close(dfd);
+
+        assert_se((dfd = open_mkdir_at_p(tfd, "a/b/c/d/e", O_CLOEXEC, 0700)) >= 0);
+        assert_se(faccessat(tfd, "a/b/c/d/e", F_OK, 0) >= 0);
+        dfd = safe_close(dfd);
```

Comment:
maybe add a test with ".." and "." and absolute paths, to check they are handled safely.